File: 04_BehaviouralDesignPatterns/03_StrategyPattern/cpp/Main.cpp
#include <iostream>
#include <string>
#include <memory>

// Strategy interface
class MatchingStrategy {
public:
    virtual void matchRider(const std::string& riderLocation) = 0;
    virtual ~MatchingStrategy() = default;
};

// Concrete strategies
class NearestDriverStrategy : public MatchingStrategy {
public:
    void matchRider(const std::string& riderLocation) override {
        std::cout << "Matching rider at " << riderLocation << " with nearest driver." << std::endl;
    }
};

class FastestDriverStrategy : public MatchingStrategy {
public:
    void matchRider(const std::string& riderLocation) override {
        std::cout << "Matching rider at " << riderLocation << " with fastest driver." << std::endl;
    }
};

class CheapestDriverStrategy : public MatchingStrategy {
public:
    void matchRider(const std::string& riderLocation) override {
        std::cout << "Matching rider at " << riderLocation << " with cheapest driver." << std::endl;
    }
};

// Context
class RideMatchingService {
private:
    std::unique_ptr<MatchingStrategy> strategy;

public:
    void setMatchingStrategy(std::unique_ptr<MatchingStrategy> newStrategy) {
        strategy = std::move(newStrategy);
    }

    void matchRider(const std::string& riderLocation) {
        if (strategy) {
            strategy->matchRider(riderLocation);
        } else {
            std::cout << "No matching strategy set!" << std::endl;
        }
    }
};

int main() {
    RideMatchingService rideMatchingService;

    // Using nearest driver strategy
    rideMatchingService.setMatchingStrategy(std::make_unique<NearestDriverStrategy>());
    rideMatchingService.matchRider("Downtown");

    // Switch to fastest driver strategy
    rideMatchingService.setMatchingStrategy(std::make_unique<FastestDriverStrategy>());
    rideMatchingService.matchRider("Uptown");

    // Switch to cheapest driver strategy
    rideMatchingService.setMatchingStrategy(std::make_unique<CheapestDriverStrategy>());
    rideMatchingService.matchRider("Suburbs");

    return 0;
}

File: 04_BehaviouralDesignPatterns/03_StrategyPattern/java/Main.java
interface RideMatchingStrategy {
    void matchRider(String riderLocation);
}

class NearestDriverStrategy implements RideMatchingStrategy {
    @Override
    public void matchRider(String riderLocation) {
        System.out.println("Matching rider at " + riderLocation + " with nearest driver.");
    }
}

class FastestDriverStrategy implements RideMatchingStrategy {
    @Override
    public void matchRider(String riderLocation) {
        System.out.println("Matching rider at " + riderLocation + " with fastest driver.");
    }
}

class CheapestDriverStrategy implements RideMatchingStrategy {
    @Override
    public void matchRider(String riderLocation) {
        System.out.println("Matching rider at " + riderLocation + " with cheapest driver.");
    }
}

class RideMatchingService {
    private RideMatchingStrategy matchingStrategy;

    public void setMatchingStrategy(RideMatchingStrategy strategy) {
        this.matchingStrategy = strategy;
    }

    public void matchRider(String riderLocation) {
        if (matchingStrategy == null) {
            System.out.println("Please set a matching strategy first.");
            return;
        }
        matchingStrategy.matchRider(riderLocation);
    }
}

public class Main {
    public static void main(String[] args) {
        RideMatchingService rideMatchingService = new RideMatchingService();

        // Using nearest driver strategy
        rideMatchingService.setMatchingStrategy(new NearestDriverStrategy());
        rideMatchingService.matchRider("Downtown");

        // Using fastest driver strategy
        rideMatchingService.setMatchingStrategy(new FastestDriverStrategy());
        rideMatchingService.matchRider("Uptown");

        // Using cheapest driver strategy
        rideMatchingService.setMatchingStrategy(new CheapestDriverStrategy());
        rideMatchingService.matchRider("Suburbs");
    }
}